File: src/python/meshioplusplus/_data_common.py
"""Shared vocabulary for the *data* operations.

These operations act on a mesh's ``point_data`` / ``cell_data`` / ``field_data``
arrays rather than on its geometry, which they never modify. A dependency-free
mesh *operation* (not a file format), so nothing here is registered in the
format registry.

Non-finite values (NaN, +/-inf) follow one policy across every data operation:
they are **always excluded from reductions** (min/max/mean/stddev and the
averaging accumulators), and ``nan_policy`` only decides what reaches the
output -- ``"ignore"`` passes them through, ``"replace"`` substitutes
``nan_replacement``, and ``"fail"`` raises. ``data_info`` never raises; it
counts them instead.

Public API:
    normalize_location -- canonicalize a location name
    location_map       -- the mesh attribute holding a location's arrays
    available_keys     -- sorted names at a location
    require_key        -- raise a listing error if a key is absent
"""

from __future__ import annotations

_LOCATION_ALIASES = {
    "point": "point_data",
    "point_data": "point_data",
    "node": "point_data",
    "cell": "cell_data",
    "cell_data": "cell_data",
    "element": "cell_data",
    "field": "field_data",
    "field_data": "field_data",
}

LOCATIONS = ("point_data", "cell_data", "field_data")


def normalize_location(location: str) -> str:
    """Return the canonical ``*_data`` name for ``location``."""
    try:
        return _LOCATION_ALIASES[location]
    except KeyError:
        raise ValueError(
            f"meshio++: unknown data location '{location}' "
            "(expected 'point', 'cell' or 'field')"
        ) from None


def location_map(mesh, location: str) -> dict:
    """Return the mesh dict holding ``location``'s arrays."""
    return getattr(mesh, normalize_location(location))


def available_keys(mesh, location: str) -> list:
    """Return the sorted array names present at ``location``."""
    return sorted(location_map(mesh, location).keys())


def require_key(mesh, location: str, name: str) -> None:
    """Raise ``ValueError`` naming the available keys if ``name`` is absent.

    ``ValueError`` rather than the more idiomatic ``KeyError`` so that both code
    paths agree: the C++ core raises ``std::invalid_argument``, which pybind11
    surfaces as ``ValueError``. Windows CI runs entirely on the numpy fallbacks,
    so a mismatch here would be a platform-dependent test failure.
    """
    loc = normalize_location(location)
    if name in location_map(mesh, loc):
        return
    keys = available_keys(mesh, loc)
    detail = f"available: {', '.join(keys)}" if keys else f"the mesh has no {loc}"
    raise ValueError(f"meshio++: no {loc} array named '{name}' ({detail})")


def num_components(array) -> int:
    """Return the product of ``array``'s trailing dimensions (1 for a scalar)."""
    import numpy as np

    a = np.asarray(array)
    if a.ndim < 2:
        return 1
    n = 1
    for d in a.shape[1:]:
        n *= int(d)
    return max(n, 1)

File: src/cpp/src/operations/data_common.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// Shared vocabulary for the data operations: the DataLocation / NanPolicy
// string parsers used by both CLIs and the WASM binding, the sorted name
// listing, and the "unknown key (available: ...)" diagnostic every data
// operation reports. See operations/data_common.hpp for the contract.

// System includes
#include <cstddef>
#include <stdexcept>
#include <string>
#include <vector>

// Project includes
#include "meshioplusplus/operations/data_common.hpp"

namespace meshioplusplus {

DataLocation data_location_from_name(const std::string& rName) {
    if (rName == "point" || rName == "point_data" || rName == "node")
        return DataLocation::Point;
    if (rName == "cell" || rName == "cell_data" || rName == "element")
        return DataLocation::Cell;
    if (rName == "field" || rName == "field_data")
        return DataLocation::Field;
    throw std::invalid_argument("meshio++: unknown data location '" + rName +
                                "' (expected 'point', 'cell' or 'field')");
}

const char* data_location_name(DataLocation Location) {
    switch (Location) {
        case DataLocation::Point:
            return "point_data";
        case DataLocation::Cell:
            return "cell_data";
        case DataLocation::Field:
            return "field_data";
    }
    return "point_data";  // unreachable
}

NanPolicy nan_policy_from_name(const std::string& rName) {
    if (rName == "ignore")
        return NanPolicy::Ignore;
    if (rName == "replace")
        return NanPolicy::Replace;
    if (rName == "fail")
        return NanPolicy::Fail;
    throw std::invalid_argument("meshio++: unknown NaN policy '" + rName +
                                "' (expected 'ignore', 'replace' or 'fail')");
}

std::vector<std::string> data_names(const Mesh& rMesh, DataLocation Location) {
    switch (Location) {
        case DataLocation::Point:
            return rMesh.PointDataNames();
        case DataLocation::Cell:
            return rMesh.CellDataNames();
        case DataLocation::Field:
            return rMesh.FieldDataNames();
    }
    return {};  // unreachable
}

bool data_has(const Mesh& rMesh, DataLocation Location, const std::string& rName) {
    switch (Location) {
        case DataLocation::Point:
            return rMesh.HasPointData(rName);
        case DataLocation::Cell:
            return rMesh.HasCellData(rName);
        case DataLocation::Field:
            return rMesh.HasFieldData(rName);
    }
    return false;  // unreachable
}

std::string data_unknown_key_message(const Mesh& rMesh, DataLocation Location,
                                     const std::string& rName) {
    const char* loc = data_location_name(Location);
    const std::vector<std::string> available = data_names(rMesh, Location);
    std::string msg = "meshio++: no ";
    msg += loc;
    msg += " array named '" + rName + "' (";
    if (available.empty()) {
        msg += "the mesh has no ";
        msg += loc;
    } else {
        msg += "available: ";
        for (std::size_t i = 0; i < available.size(); ++i) {
            if (i != 0)
                msg += ", ";
            msg += available[i];
        }
    }
    msg += ")";
    return msg;
}

std::size_t data_num_components(const NDArray& rArray) {
    const std::vector<std::size_t>& shape = rArray.Shape();
    if (shape.size() < 2)
        return 1;
    std::size_t n = 1;
    for (std::size_t d = 1; d < shape.size(); ++d)
        n *= shape[d];
    return n == 0 ? 1 : n;
}

}  // namespace meshioplusplus
